Add devcontainer readme (#481)
NVIDIA/cccl: .devcontainer/README.md

    ```bash
    git clone https://github.com/nvidia/cccl.git
    cd cccl
    ./.devcontainer/launch.sh --docker
    ```
    This script starts an interactive shell as the `coder` user inside the container with the local `cccl/` directory mirrored into `/home/coder/cccl`.

    For specific environments, use the `--cuda` and `--host` options:
    ```bassh
    ./.devcontainer/launch.sh --docker --cuda 12.2 --host gcc10
    ```
    See `./.devcontainer/launch.sh --help` for more information.

2. Done. See the [contributing guide](../CONTRIBUTING.md#building-and-testing) for instructions on how to build and run tests.

## Available Environments

CCCL provides environments for both the oldest and newest supported CUDA versions with all compatible host compilers.

Look in the [`.devcontainer/`](.) directory to see the available configurations.The top-level [`devcontainer.json`](./devcontainer.json) serves as the default environment. All `devcontainer.json` files in the `cuda<CTK_VERSION>-<HOST-COMPILER>` sub-directories are derived from this top-level file, just with different base images for the different CUDA and host compiler versions.

## VSCode Customization

> By default, CCCL's Dev Containers come with certain VSCode settings and extensions configured by default as can be seen in the [`devcontainer.json`](./devcontainer.json) file. This can be further customized by users without needing to modify the `devcontainer.json` file directly.

Reviewer: Suggestion: consider using ` - ` or `, ` to avoid a run-on sentance.
```suggestion
By default, CCCL's Dev Containers come with certain VSCode settings and extensions configured by default, as can be seen in the [`devcontainer.json`](./devcontainer.json) file. This can be further customized by users without needing to modify the `devcontainer.json` file directly.
```

For extensions, the [`dev.containers.defaultExtensions` setting](https://code.visualstudio.com/docs/devcontainers/containers#_always-installed-extensions) allows listing extensions that will always be installed.

For more general customizations, VSCode allows using a dotfile repository. See the [VSCode documentation](https://code.visualstudio.com/docs/devcontainers/containers#_personalizing-with-dotfile-repositories) for more information.

## GitHub Codespaces

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/NVIDIA/cccl?quickstart=1&devcontainer_path=.devcontainer%2Fdevcontainer.json)

> One of the benefits of Dev Containers is that they integrate natively with [GitHub Codespaces](https://github.com/features/codespaces). Codespaces provide a VSCode development environment right in your browser running on a machine in the cloud. This provides a truly one-click, turnkey development environment where you can build tests with no other setup required.

Reviewer: ```suggestion
One of the benefits of Dev Containers is that they integrate natively with [GitHub Codespaces](https://github.com/features/codespaces). Codespaces provide a VSCode development environment right in your browser running on a machine in the cloud. This provides a truly one-click, turnkey development environment where you can develop, build, and test with no other setup required.
```

> **Note**
> Platform Note: The instructions in this README are tailored for Linux environments. If you're using Windows, please be patient as we're actively working on adding support and instructions for Windows development environments. Stay tuned!

# CCCL Dev Containers

To ensure consistency and ease of setup, the CUDA C++ Libraries (CCCL) team uses [Development Containers](https://code.visualstudio.com/docs/devcontainers/containers).

> This guide offers information on setting up container in Visual Studio Code and manually with Docker.

Reviewer 2: ```suggestion
This guide offers information on setting up containers in Visual Studio Code and manually with Docker.
```

## Quickstart: VSCode (Recommended)

### **Prerequisites**:
- [Visual Studio Code](https://code.visualstudio.com/)
- [Remote - Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers)

### **Step 1: Clone the Repository**

- **Easy Way**:
Copy and paste the following URI into your browser to automatically clone and open in VSCode:

    ```
    vscode://vscode.git/clone?url=https://github.com/nvidia/cccl.git
    ```

- **Manual Way**:
  ```bash
  git clone https://github.com/nvidia/cccl.git
  ```

### **Step 2: Launch Development Environment in VSCode**

- Open the cloned directory in VSCode.
- A prompt will appear in the lower right corner suggesting to "Reopen in Container". Click it.
- A list of devcontainers will be shown. These are named with a pattern `cuda<CTK_VERSION>-<HOST-COMPILER>`. Select the desired environment to start coding!

## Quickstart: Docker (Manual Approach)

### **Prerequisites**:
- [Docker](https://docs.docker.com/desktop/install/linux-install/)

### **Step 1: Clone the Repository**
```bash
git clone https://github.com/nvidia/cccl.git
```

### **Step 2: Launch Docker Container**

> Navigate to the cloned directory and use the [`launch.sh`](./launch.sh) script:

Reviewer 2: ```suggestion
Navigate to the cloned directory and use the [`launch.sh`](./.devcontainer/launch.sh) script:
```

```bash
cd cccl
./.devcontainer/launch.sh --docker
```

> This script initiates a Docker container tailored for the specified CUDA and compiler versions. Inside, you'll get an interactive shell session with the `cccl` repository already mounted, allowing direct editing and testing.

Reviewer 2: ```suggestion
This script initializes a Docker container tailored for the specified CUDA and compiler versions. Inside, you'll get an interactive shell session with the `cccl` repository already mounted, allowing direct editing and testing.
```

> For a deeper dive into dev containers, configurations, and benefits, continue reading below.

Reviewer 2: I think it's capitalized everywhere else.
```suggestion
For a deeper dive into Dev Containers, configurations, and benefits, continue reading below.
```

## Overview

> The CUDA C++ Libraries (CCCL) team uses [Development Containers, or "Dev Containers"](https://code.visualstudio.com/docs/devcontainers/containers), to provide a consistent, reproducible, and isolated development environment shared for both local development and CI.

Reviewer 2: Maybe better to link to https://containers.dev/ since you're describing the concept and not just the VSCode extension.

> The CUDA C++ Libraries (CCCL) team uses [Development Containers, or "Dev Containers"](https://code.visualstudio.com/docs/devcontainers/containers), to provide a consistent, reproducible, and isolated development environment shared for both local development and CI.

Reviewer 3: Missing a "C" again.
```suggestion
The CUDA C++ Core Libraries (CCCL) team uses [Development Containers, or "Dev Containers"](https://code.visualstudio.com/docs/devcontainers/containers), to provide a consistent, reproducible, and isolated development environment shared for both local development and CI.
```

We encourage external contributors to use them as they provide an effortless onramp to setting up your environment for contributions. With just a few actions in Visual Studio Code or a basic Docker command, you're ready in the same environment the CCCL team works in.

> If Docker feels daunting, Dev Containers make it more intuitive. They manage Docker's intricacies, freeing you to code. No need for complex Docker commands or profound containerization knowledge. The `devcontainer.json` file dictates each environment, designating a base Docker image with the desired CUDA Toolkit (CTK) and host compiler version. This file might also detail extra configurations, tools, or settings for further refinement.

Reviewer 3: > They manage Docker's intricacies, freeing you to code. No need for complex Docker commands or profound containerization knowledge.

I'd revise this to sound a little less magical. It's setting sky-high expectations that will feel like a major letdown if something goes awry.

Our containers are provided by images built from [`rapidsai/devcontainers`](https://github.com/rapidsai/devcontainers).

### Available Dev Container Configurations:

A frequent challenge in CUDA development is managing different CUDA toolkit and compiler versions. Dev Containers address this by presenting environments tailored to numerous CTK and compiler combinations. By default, the CCCL team provides configurations for both the oldest and newest supported CUDA versions, with every supported host compiler for those CUDA versions.

To see the configurations available, look through the [`.devcontainer` directory](.). Each environment resides in its folder, named in the format `cuda<CTK_VERSION>-<HOST-COMPILER>`, with the related `devcontainer.json` explaining its specs. All Dev Container environments are fundamentally identical, only differing in the CUDA Toolkit (CTK) and host compiler versions. This uniformity guarantees that all developers, either from the CCCL team or external contributors, work within a standardized environment, reducing discrepancies from diverse local setups.

> 

Reviewer 3: ```suggestion
```

## Using Dev Containers with VSCode

Using Development Containers in Visual Studio Code provides a seamless integration, allowing developers to work within a containerized environment directly from the editor.

> ### **1. Open the Cloned Directory in VSCode**

Reviewer 3: This is heavily repeating the quickstart. Do we need both?

Upon choosing any of the above methods, you'll be presented with a list of available devcontainers. They're named following a pattern indicating their configurations, such as `cuda<CTK_VERSION>-<HOST-COMPILER>`. Select the desired environment from this list.

![Shows list of available container environments.](./img/container_list.png)

### **4. Initialization and Workspace Integration**

When you select a Dev Container configuration, VSCode starts the Docker image, automatically pulling it if not already present. This may take a few minutes if it needs to download the container image.

Post initialization, your local `cccl` repository is mounted into the container. This setup offers an integrated workspace where changes in VSCode directly reflect inside the container and vice-versa.

### **5. GitHub Authentication**

Shortly after starting, the Dev Container will prompt you to authenticate with GitHub. While optional, this authentication grants access to NVIDIA's distributed `sccache`, significantly accelerating builds. However, this feature is currently exclusive to NVIDIA employees who are members of the `NVIDIA` or `rapidsai` GitHub organizations.

> Follow the prompt, which will have you enter a one-time code at https://github.com/login/device. Follow the instructions on the GitHub web page and after succesfully authenticating, your GitHub account is used to authenticate access to the sccache distributed storage. Succesfully completing this process looks like the following:

Reviewer 3: ```suggestion
Follow the prompt, which will have you enter a one-time code at https://github.com/login/device. Follow the instructions on the GitHub web page and after successfully authenticating, your GitHub account is used to authenticate access to the sccache distributed storage. Successfully completing this process looks like the following:
```

![Shows authentication with GitHub to access sccache bucket.](./img/github_auth.png)

This step is entirely optional and will not impact any functionality other than not being able to read or write to the distributed sccache storage. `sccache` will still write a local cache to your filesystem that will still accelerate local rebuilds.

If at any time you want to run this authentication process manually, you can do so by running the `devcontainer-utils-vault-s3-init` script inside the container.

For more information about the sccache configuration and authentication, see the documentation at [`rapidsai/devcontainers`](https://github.com/rapidsai/devcontainers/blob/branch-23.10/USAGE.md#build-caching-with-sccache).

### **6. Working Inside the Devcontainer**

With the container launched, you're now in an environment tailored for CCCL development. Whether coding, running tests, or conducting other tasks, you'll benefit from the tailored tools and settings of this environment.

> For more information on building and running tests, see the [contributer guide](../CONTRIBUTING.md).

Reviewer 3: ```suggestion
For more information on building and running tests, see the [contributor guide](../CONTRIBUTING.md).
```

## Using Docker Images Directly (Manual Approach)

While Visual Studio Code with Dev Containers offers a smooth development experience, there might be developers who lean towards a direct Docker approach. For those, we've simplified the process with our `launch.sh` script.

Before using the script, you might want to explore the available configurations.
The `.devcontainer` directory and its subdirectories house these configurations, with each subdirectory typically named as `cuda<CTK_VERSION>-<HOST-COMPILER>`.

### **Steps**:

1. **Use the `launch.sh` Script**:
   To initiate your desired development environment, run the `launch.sh` script with flags specifying your desired CUDA and compiler versions:
   ```bash
   ./launch.sh --cuda 12.2 --host gcc12 --docker
   ```
   This command pulls the appropriate Docker image, runs it, and provides an interactive shell within the container.

2. **Workspace Setup**:

>    The `launch.sh` script automatically mounts the CCCL repository directory into the container. This ensures you have a mirrored workspace inside the container, resembling your local setup.

Reviewer 4: ```suggestion
   The `launch.sh` script automatically mounts the CCCL repository directory into the container. It can be accessed in the container through the `/workspace` directory. This ensures you have a mirrored workspace inside the container, resembling your local setup.
```

3. **Working Inside the Container**:
   Inside the container, you are equipped with all tools and configurations designed for CCCL development. You can initiate builds, run tests, or perform any other development tasks as you would in a local setup.

4. **Exiting the Container**:
   After you're finished with your tasks, simply type `exit` in the interactive shell to leave. Upon exit, the container gets removed, assuring a pristine state for your next session.

## `launch.sh`: Development Container Launcher Script

### Overview

This script provides a streamlined process to launch a development container environment tailored for CCCL development.
The environment can be started either in Docker directly or within Visual Studio Code (VSCode) using its "Remote - Containers" feature.

### Usage

```bash
./launch.sh [-c|--cuda <CUDA version>] [-H|--host <Host compiler>] [-d|--docker]
```

### Options

- `-c, --cuda`: Specify the CUDA version for the environment. For example, `12.2`.
- `-H, --host`: Specify the host compiler version for the environment. For example, `gcc12`.

> - `-d, --docker`: By default, the script assumes you want to launch the environment in VSCode. Use this option if you want to run the development environment in Docker directly.

Reviewer 2: I wonder if this flag should be flipped. IIUC VSCode users will mostly be coming from the VSCode GUI, so the default `launch.sh` user is probably going to be going for raw Docker right?

Author: I'd been considering the exact same thing. The fact that you had the same idea tells me I should probably make that change :slightly_smiling_face:.

> **Note**

> > The instructions in this README are specific to Linux development environments. Instructions for Windows is coming soon!

Reviewer 5: ```suggestion
> The instructions in this README are specific to Linux development environments. Instructions for Windows are coming soon!
```
